### PR TITLE
[SDAG] Simplify `SDNodeFlags` with bitwise logic

### PR DESCRIPTION
**Pull request by dtcxzyw**
This patch allows using enumeration values directly and simplifies the implementation with bitwise logic. It addresses the comment in https://github.com/llvm/llvm-project/pull/113808#discussion_r1819923625.

### CODE REVIEW

File: llvm/include/llvm/CodeGen/SelectionDAGNodes.h [hunk 1]
```diff
@@ -378,36 +378,46 @@ template<> struct simplify_type<SDUse> {
...
+  friend class SDNode;
+
+  unsigned Flags = 0;
+
+  template <unsigned Flag> void setFlag(bool B) {
+    Flags = (Flags & ~Flag) | (B ? Flag : 0);
+  }
 
 public:
+  enum {
```
Comment (kparzysz) [inline llvm/include/llvm/CodeGen/SelectionDAGNodes.h:390]: Could you make it `enum : unsigned` (or `enum : decltype(Flags)`)?

Comment (arsenm) [inline llvm/include/llvm/CodeGen/SelectionDAGNodes.h:390]: Is it worth marking this with LLVM_DECLARE_ENUM_AS_BITMASK

**Review by antoniofrighetto — COMMENTED** [review]
Is the whole change supposed to be part of a NFC, isn't it?

**Comment by dtcxzyw** [comment]
> Is the whole change supposed to be part of a NFC, isn't it?

The second commit breaks SDNodeFlags API.

**Comment by goldsteinn** [comment]
IMO since github doesn't have a great interface for reviewing single-commits in a series, you should split the first commit to its own proper NFC PR.

File: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp [hunk 2]
```diff
@@ -2892,11 +2892,11 @@ SDValue DAGCombiner::visitADDLike(SDNode *N) {
         if (N->getFlags().hasNoUnsignedWrap() &&
             N0->getFlags().hasNoUnsignedWrap() &&
             N0.getOperand(0)->getFlags().hasNoUnsignedWrap()) {
-          Flags.setNoUnsignedWrap(true);
+          Flags = SDNodeFlags::NoUnsignedWrap;
```
Comment (goldsteinn) [inline llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:2895]: Here and below, imo just make it all `|=`.

File: llvm/include/llvm/CodeGen/SelectionDAGNodes.h [hunk 3]
```diff
@@ -1006,6 +1016,7 @@ END_TWO_BYTE_PACK()
 
   SDNodeFlags getFlags() const { return Flags; }
   void setFlags(SDNodeFlags NewFlags) { Flags = NewFlags; }
+  void clearFlags(unsigned Mask) { Flags &= ~Mask; }
```
Comment (topperc) [inline llvm/include/llvm/CodeGen/SelectionDAGNodes.h:1025]: The interface here is a little confusing now. The existence of clearFlags could make someone think that setFlags would have OR semantics.

Comment (dtcxzyw) [reply llvm/include/llvm/CodeGen/SelectionDAGNodes.h:1025]: Do you have any suggestion on the naming? `maskOutFlags`? `dropFlags`?

Comment (topperc) [reply llvm/include/llvm/CodeGen/SelectionDAGNodes.h:1025]: I'm ok with `dropFlags`

Comment (dtcxzyw) [reply llvm/include/llvm/CodeGen/SelectionDAGNodes.h]: Done.

**Review by arsenm — APPROVED** [review]

File: llvm/include/llvm/CodeGen/SelectionDAGNodes.h (hunk 3 above)
Comment (dtcxzyw) [reply llvm/include/llvm/CodeGen/SelectionDAGNodes.h]: Emm... It is likely to break some downstream implementations.

**Review by antoniofrighetto — APPROVED** [review]
LGTM, thanks.

File: llvm/include/llvm/CodeGen/SelectionDAGNodes.h (hunk 3 above)
Comment (goldsteinn) [reply llvm/include/llvm/CodeGen/SelectionDAGNodes.h]: Fair enough;

**Pull request merged**